### DISCUSSION TITLE
[IR] Add Line Number Tracing in MLIR

### DISCUSSION THREAD

**Discussion by chhzh123** [cornell-zhang/allo#224] [Projects]
The existing Allo IR builder does not include the Python source code line number in the IR, which makes it hard to trace back to the original function from MLIR to Python and debug. Therefore, we want to include this information in the IR by explicitly setting the `Location` field for each operation. See the [customize function](https://github.com/cornell-zhang/allo/blob/main/allo/customize.py#L84).

Reference:
* https://github.com/triton-lang/triton

**Comment by tmao8** [comment]
Forked repository and added code
Read paper
Looking at customize.py and other code base to try and get a starting point for tracing errors.

**Comment by chhzh123** [comment]
@tmao8 Can you add an update here? I think it is also good to create a PR after you have some test cases. For simple testing, you can just turn the MLIR module into a string and then directly compare the string.

**Comment by tmao8** [comment]
Created an optional parameter to ASTTransformer callable for file name (the file the function is from).  Using this parameter, the ASTBuilder will set a global variable file_name, and every time the ASTBuilder is called it will have the file name stored. Each node can then transform with location specified by the node's lineno and col offset. The MLIR now has line numbers from the original python frontend if printed using s.module.operation.print() and specifying enable_debug_info=True and use_local_scope=True

**Comment by chhzh123** [comment]
Done in #295
